ROOT Dictionary Error For `CUDADataFormats`

When trying to run on Run3 RAW files on which the new HLT step has been run (from 12_3_X on) some ROOT dictionary error shows up (see below for a link to a sample log) if the definition of `CUDADataFormats` (such as `TrackSoAHeterogeneousT` or `TrackingRecHit2DHeterogeneous `)  is changed. This has been spotted and affects https://github.com/cms-sw/cmssw/pull/38761. Even if not necessary (as far as I have understood) for transient classes, the usual class versioning instructions (and the `CMS_CLASS_VERSION` tool) have been tested unsuccessfully.

I've set up a basic test (simply changing `maxNumber()` in `TrackSoAHeterogeneousT`) 




```
scram p -n dev CMSSW_12_5_0_pre2 
cd dev/src/ 
cmsenv 
git cms-init 
git cms-merge-topic AdrianoDee:cudaformats_failures 
scram b -j 16 
cmsDriver.py step3 -s RAW2DIGI:RawToDigi_pixelOnly,RECO:reconstruction_pixelTrackingOnly,VALIDATION:@pixelTrackingOnlyValidation,DQM:@pixelTrackingOnlyDQM --conditions auto:phase1_2022_realistic --datatier GEN-SIM-RECO,DQMIO -n 10 --eventcontent RECOSIM,DQM --geometry DB:Extended --era Run3 --procModifiers pixelNtupletFit,gpu --filein /store/relval/CMSSW_12_5_0_pre2/RelValTTbar_14TeV/GEN-SIM-DIGI-RAW/PU_124X_mcRun3_2022_realistic_v3-v1/2580000/148e5f22-ea2b-4dfb-b64f-79e50be63665.root --no_exec --python_filename=step3.py 
cmsRun step3.py
```

that outputs the dictionary error

```
[adrianodif@tesla04 test_class]$ cmsRun step3.py 
%MSG-i CUDAService:  (NoModuleName) 18-Jul-2022 10:10:14 CEST pre-events
CUDA runtime version 11.5, driver version 11.2, NVIDIA driver version 460.27.04
CUDA device 0: Tesla P100-PCIE-12GB (sm_60)
%MSG
18-Jul-2022 10:11:22 CEST  Initiating request to open file file:/cms//store/relval/CMSSW_12_5_0_pre2/RelValTTbar_14TeV/GEN-SIM-DIGI-RAW/PU_124X_mcRun3_2022_realistic_v3-v1/2580000/148e5f22-ea2b-4dfb-b64f-79e50be63665.root
18-Jul-2022 10:11:23 CEST  Initiating request to open file root://xrootd-cms.infn.it:1194//store/relval/CMSSW_12_5_0_pre2/RelValTTbar_14TeV/GEN-SIM-DIGI-RAW/PU_124X_mcRun3_2022_realistic_v3-v1/2580000/148e5f22-ea2b-4dfb-b64f-79e50be63665.root
%MSG-w XrdAdaptorInternal:  file_open 18-Jul-2022 10:11:31 CEST pre-events
Failed to open file at URL root://grid143.kfki.hu:11000//store/relval/CMSSW_12_5_0_pre2/RelValTTbar_14TeV/GEN-SIM-DIGI-RAW/PU_124X_mcRun3_2022_realistic_v3-v1/2580000/148e5f22-ea2b-4dfb-b64f-79e50be63665.root?tried=+1313xrootd-cms-redir-int.cr.cnaf.infn.it&xrdcl.requuid=356c58a3-4362-4efa-879f-8a8db6def6ae.
%MSG
%MSG-w XrdAdaptorInternal:  file_open 18-Jul-2022 10:11:32 CEST pre-events
Failed to open file at URL root://grid143.kfki.hu:11000//store/relval/CMSSW_12_5_0_pre2/RelValTTbar_14TeV/GEN-SIM-DIGI-RAW/PU_124X_mcRun3_2022_realistic_v3-v1/2580000/148e5f22-ea2b-4dfb-b64f-79e50be63665.root?tried=+1313xrootd-cms-redir-int.cr.cnaf.infn.it,&xrdcl.requuid=507d224f-cb1f-43ca-85cd-0ec2b6b98ab4.
%MSG
18-Jul-2022 10:11:32 CEST  Initiating request to open file root://cms-xrd-global.cern.ch//store/relval/CMSSW_12_5_0_pre2/RelValTTbar_14TeV/GEN-SIM-DIGI-RAW/PU_124X_mcRun3_2022_realistic_v3-v1/2580000/148e5f22-ea2b-4dfb-b64f-79e50be63665.root
%MSG-w XrdAdaptor:  file_open 18-Jul-2022 10:11:36 CEST pre-events
Data is served from cern.ch instead of original site CERN-CMSGLOBAL-02
%MSG
18-Jul-2022 10:11:43 CEST  Successfully opened file root://cms-xrd-global.cern.ch//store/relval/CMSSW_12_5_0_pre2/RelValTTbar_14TeV/GEN-SIM-DIGI-RAW/PU_124X_mcRun3_2022_realistic_v3-v1/2580000/148e5f22-ea2b-4dfb-b64f-79e50be63665.root
In file included from CUDADataFormatsTrack_xr dictionary payload:65:
In file included from /cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_0_pre2/src/CUDADataFormats/Common/interface/Product.h:6:
In file included from /cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_0_pre2/src/CUDADataFormats/Common/interface/ProductBase.h:7:
/cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_0_pre2/src/HeterogeneousCore/CUDAUtilities/interface/SharedStreamPtr.h:7:10: remark: could not acquire lock file for module 'cuda': failed to create unique file /cvmfs/cms.cern.ch/slc7_amd64_gcc10/lcg/root/6.24.07-60321294cf9ce0aa5923969e44204616/lib/cuda.pcm.lock-2277eaf6: Read-only file system [-Rmodule-build]
#include <cuda_runtime.h>
         ^
/cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_0_pre2/src/HeterogeneousCore/CUDAUtilities/interface/SharedStreamPtr.h:7:10: remark: building module 'cuda' as '/cvmfs/cms.cern.ch/slc7_amd64_gcc10/lcg/root/6.24.07-60321294cf9ce0aa5923969e44204616/lib/cuda.pcm' [-Rmodule-build]
error: unable to open output file '/cvmfs/cms.cern.ch/slc7_amd64_gcc10/lcg/root/6.24.07-60321294cf9ce0aa5923969e44204616/lib/cuda.pcm': 'No such file or directory'
/cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_0_pre2/src/HeterogeneousCore/CUDAUtilities/interface/SharedStreamPtr.h:7:10: remark: finished building module 'cuda' [-Rmodule-build]
/cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_0_pre2/src/HeterogeneousCore/CUDAUtilities/interface/SharedStreamPtr.h:7:10: fatal error: could not build module 'cuda'
#include <cuda_runtime.h>
 ~~~~~~~~^
18-Jul-2022 10:12:05 CEST  Closed file root://cms-xrd-global.cern.ch//store/relval/CMSSW_12_5_0_pre2/RelValTTbar_14TeV/GEN-SIM-DIGI-RAW/PU_124X_mcRun3_2022_realistic_v3-v1/2580000/148e5f22-ea2b-4dfb-b64f-79e50be63665.root
----- Begin Fatal Exception 18-Jul-2022 10:12:05 CEST-----------------------
An exception of category 'FatalRootError' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing input source of type PoolSource
   [2] While initializing meta data for branch: 32768TrackSoAHeterogeneousTHeterogeneousSoA_hltPixelTracksCPU__HLT.
   Additional Info:
      [a] Fatal Root Error: @SUB=TInterpreter::AutoParse
Error parsing payload code for class HeterogeneousSoA with content:

#line 1 "CUDADataFormatsTrack_xr dictionary payload"

#ifndef CMS_DICT_IMPL
  #define CMS_DICT_IMPL 1
#endif
#ifndef _REENTRANT
  #define _REENTRANT 1
#endif
#ifndef GNUSOURCE
  #define GNUSOURCE 1
#endif
#ifndef __STRICT_ANSI__
  #define __STRICT_ANSI__ 1
#endif
#ifndef GNU_GCC
  #define GNU_GCC 1
#endif
#ifndef _GNU_SOURCE
  #define _GNU_SOURCE 1
#endif
#ifndef EIGEN_DONT_PARALLELIZE
  #define EIGEN_DONT_PARALLELIZE 1
#endif
#ifndef TBB_USE_GLIBCXX_VERSION
  #define TBB_USE_GLIBCXX_VERSION 100300
#endif
#ifndef TBB_SUPPRESS_DEPRECATED_MESSAGES
  #define TBB_SUPPRESS_DEPRECATED_MESSAGES 1
#endif
#ifndef TBB_PREVIEW_RESUMABLE_TASKS
  #define TBB_PREVIEW_RESUMABLE_TASKS 1
#endif
#ifndef TBB_PREVIEW_TASK_GROUP_EXTENSIONS
  #define TBB_PREVIEW_TASK_GROUP_EXTENSIONS 1
#endif
#ifndef BOOST_SPIRIT_THREADSAFE
  #define BOOST_SPIRIT_THREADSAFE 1
#endif
#ifndef PHOENIX_THREADSAFE
  #define PHOENIX_THREADSAFE 1
#endif
#ifndef BOOST_MATH_DISABLE_STD_FPCLASSIFY
  #define BOOST_MATH_DISABLE_STD_FPCLASSIFY 1
#endif
#ifndef BOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX
  #define BOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX 1
#endif
#ifndef CMSSW_GIT_HASH
  #define CMSSW_GIT_HASH "CMSSW_12_5_0_pre2"
#endif
#ifndef PROJECT_NAME
  #define PROJECT_NAME "CMSSW"
#endif
#ifndef PROJECT_VERSION
  #define PROJECT_VERSION "CMSSW_12_5_0_pre2"
#endif
#ifndef CMSSW_REFLEX_DICT
  #define CMSSW_REFLEX_DICT 1
#endif

#define _BACKWARD_BACKWARD_WARNING_H
// Inline headers
#ifndef CUDADataFormats_Track_src_classes_h
#define CUDADataFormats_Track_src_classes_h

#include "CUDADataFormats/Common/interface/Product.h"
#include "CUDADataFormats/Common/interface/HostProduct.h"
#include "CUDADataFormats/Track/interface/TrackSoAHeterogeneousT.h"
#include "DataFormats/Common/interface/Wrapper.h"

#endif  // CUDADataFormats_Track_src_classes_h

#undef  _BACKWARD_BACKWARD_WARNING_H


----- End Fatal Exception -------------------------------------------------

```


this happening with or without `CMS_CLASS_VERSION`. At the same time something like `runTheMatrix.py -l 11634.502 -w upgrade -j 8` runs smoothly and this points to a problem in reading from already produced samples.

--- a/DataFormats/Common/src/setIsMergeable.cc
+++ b/DataFormats/Common/src/setIsMergeable.cc
@@ -17,7 +17,8 @@ namespace edm {
     // data member can only be true for run or lumi products.
     // It defaults to false. Also if it is true that means it
     // was already set.
-    if (desc.branchType() == InRun || desc.branchType() == InLumi) {
+    // Set it only for branches that are present
+    if (desc.present() and (desc.branchType() == InRun or desc.branchType() == InLumi)) {
       if (!desc.isMergeable()) {
         TClass* wrapperBaseTClass = TypeWithDict::byName("edm::WrapperBase").getClass();
         TClass* tClass = desc.wrappedType().getClass();

--- a/DataFormats/Provenance/interface/BranchDescription.h
+++ b/DataFormats/Provenance/interface/BranchDescription.h
@@ -158,9 +158,11 @@ namespace edm {
       BranchID switchAliasForBranchID_;
 
       // A TypeWithDict object for the wrapped object
+      // This is set if and only if the dropped_ is false
       TypeWithDict wrappedType_;
 
       // A TypeWithDict object for the unwrapped object
+      // This is set if and only if the dropped_ is false
       TypeWithDict unwrappedType_;
 
       // The split level of the branch, as marked

--- a/DataFormats/TestObjects/interface/ToyProducts.h
+++ b/DataFormats/TestObjects/interface/ToyProducts.h
@@ -63,6 +63,15 @@ namespace edmtest {
     ProductWithNoDictionary dummy;
   };
 
+  template <int TAG>
+  struct TransientIntParentT {
+    explicit TransientIntParentT(int i = 0) : value(i) {}
+
+    cms_int32_t value;
+  };
+  constexpr int TransientIntParentTag = 1;
+  using TransientIntParent = TransientIntParentT<TransientIntParentTag>;
+
   struct Int16_tProduct {
     explicit Int16_tProduct(int16_t i = 0, uint16_t j = 1) : value(i), uvalue(j) {}
     ~Int16_tProduct() {}

--- a/DataFormats/TestObjects/test/BuildFile_extra.xml
+++ b/DataFormats/TestObjects/test/BuildFile_extra.xml
@@ -1,0 +1,9 @@
+<library file="" name="DataFormatsTestObjectsParent1">
+  <flags LCG_DICT_HEADER="classes.h"/>
+  <flags LCG_DICT_XML="classes_def_parent1.xml"/>
+</library>
+
+<library file="" name="DataFormatsTestObjectsParent2">
+  <flags LCG_DICT_HEADER="classes.h"/>
+  <flags LCG_DICT_XML="classes_def_parent2.xml"/>
+</library>

--- a/DataFormats/TestObjects/test/classes.h
+++ b/DataFormats/TestObjects/test/classes.h
@@ -1,0 +1,2 @@
+#include "DataFormats/Common/interface/Wrapper.h"
+#include "DataFormats/TestObjects/interface/ToyProducts.h"

--- a/DataFormats/TestObjects/test/classes_def_parent1.xml
+++ b/DataFormats/TestObjects/test/classes_def_parent1.xml
@@ -1,0 +1,4 @@
+<lcgdict>
+ <class name="edmtest::TransientIntParentT<1>" persistent="false"/>
+ <class name="edm::Wrapper<edmtest::TransientIntParentT<1>>" persistent="false"/>
+</lcgdict>

--- a/DataFormats/TestObjects/test/classes_def_parent2.xml
+++ b/DataFormats/TestObjects/test/classes_def_parent2.xml
@@ -1,0 +1,4 @@
+<lcgdict>
+ <class name="edmtest::TransientIntParentT<2>" persistent="false"/>
+ <class name="edm::Wrapper<edmtest::TransientIntParentT<2>>" persistent="false"/>
+</lcgdict>

--- a/FWCore/Framework/interface/stream/ThinningProducer.h
+++ b/FWCore/Framework/interface/stream/ThinningProducer.h
@@ -160,6 +160,12 @@ namespace edm {
     ProductRegistry::ProductList const& productList = productRegistry.productList();
     for (auto const& product : productList) {
       BranchDescription const& desc = product.second;
+      if (desc.dropped()) {
+        // Dropped branch does not have type information, but they can
+        // be ignored here because all of the parent/thinned/association
+        // branches are expected to be present
+        continue;
+      }
       if (desc.unwrappedType().typeInfo() == typeid(Collection)) {
         if (desc.produced() && desc.moduleLabel() == moduleDescription().moduleLabel() &&
             desc.productInstanceName().empty()) {

--- a/FWCore/Framework/test/stubs/ToyIntProducers.cc
+++ b/FWCore/Framework/test/stubs/ToyIntProducers.cc
@@ -352,6 +352,46 @@ namespace edmtest {
 
   //--------------------------------------------------------------------
   //
+  // Produces a TransientIntParent instance.
+  //
+  class TransientIntParentProducer : public edm::global::EDProducer<> {
+  public:
+    explicit TransientIntParentProducer(edm::ParameterSet const& p)
+        : token_{produces<TransientIntParent>()}, value_(p.getParameter<int>("ivalue")) {}
+    void produce(edm::StreamID, edm::Event& e, edm::EventSetup const& c) const override;
+
+  private:
+    const edm::EDPutTokenT<TransientIntParent> token_;
+    const int value_;
+  };
+
+  void TransientIntParentProducer::produce(edm::StreamID, edm::Event& e, edm::EventSetup const&) const {
+    // EventSetup is not used.
+    e.emplace(token_, value_);
+  }
+
+  //--------------------------------------------------------------------
+  //
+  // Produces a IntProduct instance from a TransientIntParent
+  //
+  class IntProducerFromTransientParent : public edm::global::EDProducer<> {
+  public:
+    explicit IntProducerFromTransientParent(edm::ParameterSet const& p)
+        : putToken_{produces<IntProduct>()}, getToken_{consumes(p.getParameter<edm::InputTag>("src"))} {}
+    void produce(edm::StreamID, edm::Event& e, edm::EventSetup const& c) const override;
+
+  private:
+    const edm::EDPutTokenT<IntProduct> putToken_;
+    const edm::EDGetTokenT<TransientIntParent> getToken_;
+  };
+
+  void IntProducerFromTransientParent::produce(edm::StreamID, edm::Event& e, edm::EventSetup const&) const {
+    // EventSetup is not used.
+    e.emplace(putToken_, e.get(getToken_).value);
+  }
+
+  //--------------------------------------------------------------------
+  //
   // Produces an Int16_tProduct instance.
   //
   class Int16_tProducer : public edm::global::EDProducer<> {
@@ -907,6 +947,8 @@ DEFINE_FWK_MODULE(ConsumingIntProducer);
 DEFINE_FWK_MODULE(EventNumberIntProducer);
 DEFINE_FWK_MODULE(TransientIntProducer);
 DEFINE_FWK_MODULE(IntProducerFromTransient);
+DEFINE_FWK_MODULE(edmtest::TransientIntParentProducer);
+DEFINE_FWK_MODULE(edmtest::IntProducerFromTransientParent);
 DEFINE_FWK_MODULE(Int16_tProducer);
 DEFINE_FWK_MODULE(AddIntsProducer);
 DEFINE_FWK_MODULE(AddAllIntsProducer);

--- a/IOPool/Input/src/PoolSource.cc
+++ b/IOPool/Input/src/PoolSource.cc
@@ -121,6 +121,10 @@ namespace edm {
           //now make sure this is marked as not dropped else the product will not be 'get'table from the Event
           auto itFound = fullList.find(item.first);
           if (itFound != fullList.end()) {
+            // If the branch in primary file was dropped, need to initilize the dictionary information
+            if (itFound->second.dropped()) {
+              itFound->second.initFromDictionary();
+            }
             itFound->second.setDropped(false);
           }
         }

--- a/IOPool/Input/src/RootFile.cc
+++ b/IOPool/Input/src/RootFile.cc
@@ -1898,7 +1898,11 @@ namespace edm {
     ProductRegistry::ProductList& pList = inputProdDescReg.productListUpdator();
     for (auto& product : pList) {
       BranchDescription& prod = product.second;
-      prod.init();
+      // Initialize BranchDescription from dictionary only if the
+      // branch is present. This allows a subsequent job to process
+      // data where a dictionary of a transient parent branch has been
+      // removed from the release after the file has been written.
+      prod.initBranchName();
       if (prod.branchType() == InProcess) {
         std::vector<std::string> const& processes = storedProcessBlockHelper.processesWithProcessBlockProducts();
         auto it = std::find(processes.begin(), processes.end(), prod.processName());
@@ -1912,6 +1916,9 @@ namespace edm {
         }
       } else {
         treePointers_[prod.branchType()]->setPresence(prod, newBranchToOldBranch(prod.branchName()));
+      }
+      if (prod.present()) {
+        prod.initFromDictionary();
       }
     }
   }
@@ -2053,7 +2060,7 @@ namespace edm {
       TString tString;
       for (ProductRegistry::ProductList::iterator it = prodList.begin(), itEnd = prodList.end(); it != itEnd;) {
         BranchDescription const& prod = it->second;
-        if (prod.branchType() != InEvent && prod.branchType() != InProcess) {
+        if (prod.present() and prod.branchType() != InEvent and prod.branchType() != InProcess) {
           TClass* cp = prod.wrappedType().getClass();
           void* p = cp->New();
           int offset = cp->GetBaseClassOffset(edProductClass_);

--- a/IOPool/Input/test/BuildFile.xml
+++ b/IOPool/Input/test/BuildFile.xml
@@ -11,4 +11,5 @@
   </library>
  
  <test name="TestIOPoolInputRepeating" command="testRepeatingCachedRootSource.sh"/>
+ <test name="TestIOPoolInputNoParentDictionary" command="testNoParentDictionary.sh"/>
 </environment>

--- a/IOPool/Input/test/PoolNoParentDictionaryTestStep1_cfg.py
+++ b/IOPool/Input/test/PoolNoParentDictionaryTestStep1_cfg.py
@@ -1,0 +1,24 @@
+import FWCore.ParameterSet.Config as cms
+
+process = cms.Process("TESTPROD")
+process.maxEvents.input = 11
+
+process.source = cms.Source("EmptySource",
+    firstLuminosityBlock = cms.untracked.uint32(6),
+    numberEventsInLuminosityBlock = cms.untracked.uint32(3),
+    firstRun = cms.untracked.uint32(561),
+    numberEventsInRun = cms.untracked.uint32(7)
+)
+
+process.parentIntProduct = cms.EDProducer("edmtest::TransientIntParentProducer", ivalue = cms.int32(1))
+
+process.intProduct = cms.EDProducer("edmtest::IntProducerFromTransientParent",
+    src = cms.InputTag("parentIntProduct")
+)
+
+process.output = cms.OutputModule("PoolOutputModule",
+    fileName = cms.untracked.string('noparentdict_step1.root')
+)
+
+process.p = cms.Path(process.parentIntProduct+process.intProduct)
+process.ep = cms.EndPath(process.output)

--- a/IOPool/Input/test/PoolNoParentDictionaryTestStep2_cfg.py
+++ b/IOPool/Input/test/PoolNoParentDictionaryTestStep2_cfg.py
@@ -1,0 +1,17 @@
+import FWCore.ParameterSet.Config as cms
+
+process = cms.Process("TESTANALYSIS")
+
+process.maxEvents.input = -1
+
+process.source = cms.Source("PoolSource",
+    setRunNumber = cms.untracked.uint32(621),
+    fileNames = cms.untracked.vstring('file:noparentdict_step1.root')
+)
+
+process.analysis = cms.EDAnalyzer("IntTestAnalyzer",
+    valueMustMatch = cms.untracked.int32(1),
+    moduleLabel = cms.untracked.InputTag("intProduct"),
+)
+
+process.p = cms.Path(process.analysis)

--- a/IOPool/Input/test/testNoParentDictionary.sh
+++ b/IOPool/Input/test/testNoParentDictionary.sh
@@ -1,0 +1,51 @@
+#!/bin/bash -e
+
+function die { echo $1: status $2 ;  exit $2; }
+
+SCRAM_TEST_NAME=TestIOPoolInputNoParentDictionary
+rm -rf $SCRAM_TEST_NAME
+mkdir $SCRAM_TEST_NAME
+cd $SCRAM_TEST_NAME
+
+# Create a new CMSSW dev area
+OLD_CMSSW_BASE=${CMSSW_BASE}
+LD_LIBRARY_PATH_TO_OLD=$(echo $LD_LIBRARY_PATH | tr ':' '\n' | grep "^${CMSSW_BASE}/" | tr '\n' ':')
+scram -a $SCRAM_ARCH project $CMSSW_VERSION
+pushd $CMSSW_VERSION/src
+eval `scram run -sh`
+
+# Copy DataFormats/TestObjects code to be able to edit it to make ROOT header parsing to fail
+for DIR in ${OLD_CMSSW_BASE} ${CMSSW_RELEASE_BASE} ${CMSSW_FULL_RELEASE_BASE} ; do
+    if [ -d ${DIR}/src/DataFormats/TestObjects ]; then
+        mkdir DataFormats
+        cp -r ${DIR}/src/DataFormats/TestObjects DataFormats/
+        break
+    fi
+done
+if [ ! -e DataFormats/TestObjects ]; then
+    echo "Failed to copy DataFormats/TestObjects from local or release area"
+    exit 1;
+fi
+
+# Enable TransientIntParentT dictionaries
+cat DataFormats/TestObjects/test/BuildFile_extra.xml >> DataFormats/TestObjects/test/BuildFile.xml
+scram build -j $(nproc)
+
+popd
+
+# Add OLD_CMSSW_BASE in between CMSSW_BASE and CMSSW_RELEASE_BASE for
+# LD_LIBRARY_PATH and ROOT_INCLUDE_PATH
+export LD_LIBRARY_PATH=$(echo -n ${LD_LIBRARY_PATH} | sed -e "s|${CMSSW_BASE}/external/${SCRAM_ARCH}/lib:|${CMSSW_BASE}/external/${SCRAM_ARCH}/lib:${LD_LIBRARY_PATH_TO_OLD}:|")
+export ROOT_INCLUDE_PATH=$(echo -n ${ROOT_INCLUDE_PATH} | sed -e "s|${CMSSW_BASE}/src:|${CMSSW_BASE}/src:${OLD_CMSSW_BASE}/src:|")
+
+echo "Produce a file with TransientIntParentT<1> product"
+cmsRun ${LOCALTOP}/src/IOPool/Input/test/PoolNoParentDictionaryTestStep1_cfg.py || die 'Failed cmsRun PoolNoParentDictionaryTestStep1_cfg.py' $?
+
+
+# Then make attempt to load TransientIntParentT<1> to fail
+echo "PREVENT_HEADER_PARSING" >> ${CMSSW_BASE}/src/DataFormats/TestObjects/interface/ToyProducts.h
+rm ${CMSSW_BASE}/lib/${SCRAM_ARCH}/DataFormatsTestObjectsParent1_xr_rdict.pcm ${CMSSW_BASE}/lib/${SCRAM_ARCH}/libDataFormatsTestObjectsParent1.so
+sed -i -e 's/libDataFormatsTestObjectsParent1.so/libDataFormatsTestObjectsParent2.so/' ${CMSSW_BASE}/lib/${SCRAM_ARCH}/DataFormatsTestObjectsParent1_xr.rootmap
+
+echo "Read the file without TransientIntParentT<1> dictionary"
+cmsRun ${LOCALTOP}/src/IOPool/Input/test/PoolNoParentDictionaryTestStep2_cfg.py || die 'Failed cmsRun PoolNoParentDictionaryTestStep2_cfg.py' $?

